---
title: 'Report breakdown by guideRNA'
author: amplican
date: '`r format(Sys.time(), "%d %B %Y")`'
output:
  html_document:
    toc: true
    theme: paper
    toc_float: true
    number_sections: true
vignette: >
  %\VignetteIndexEntry{example guide report}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

```{r load data, message=F, warning=FALSE, include=FALSE}
results_folder = system.file('extdata', 'results', package = 'amplican')
library(amplican)
library(ggplot2)
alignments <- read.csv(file.path(results_folder, 'alignments_events.csv'), stringsAsFactors = FALSE)
config <- read.csv(file.path(results_folder, 'config_summary.csv'), stringsAsFactors = FALSE)
```

***

# Description  

***

**Read distribution plot** - plot shows number of reads assigned during read grouping  
**PRIMER DIMER filter** - plot shows percentage of assigned reads that have been recognized as PRIMER DIMERS  
**Cutting rates** - plot gives overview of percentage of reads (not filtered as PRIMER DIMER) that have cut  
**Frameshift** - plot shows what percentage of reads that have frameshift  
**Read heterogeneity plot** - shows what is the share of each of the unique reads in total count of all reads. The more yellow each row, the less heterogeneity in the reads, more black means reads don't repeat often and are unique  


***

# guideRNA Summary  

***

## Read distribution  

```{r plot total reads, echo=FALSE, fig.height=4, fig.width=14, message=F, warning=FALSE}
ggplot(data = config, aes(x = as.factor(guideRNA), y = log10(Reads + 1), order = guideRNA, fill = guideRNA)) +
  geom_boxplot() +
  ylab('Number of reads + 1, log10 scaled')  +
  xlab('guideRNA') + 
  theme(legend.position = 'none',
        axis.text = element_text(size = 12),
        axis.title = element_text(size = 14, face = 'bold')) +
  coord_flip()
```

## PRIMER DIMER filter  

```{r plot PRIMER DIMER percentage, echo=FALSE, fig.height=4, fig.width=14, message=F, warning=FALSE}
config$PD_percentage <- config$PRIMER_DIMER * 100/config$Reads
config$PD_percentage[is.nan(config$PD_percentage)] <- 0

ggplot(data = config, aes(x = as.factor(guideRNA), y = PD_percentage, order = guideRNA, fill = guideRNA)) +
  geom_boxplot() +
  xlab('guideRNA') + 
  ylab('Percentage of reads marked as PRIMER DIMERS')  +
  theme(axis.text = element_text(size=12),
        axis.title = element_text(size=14, face = 'bold'),
        legend.position = 'none') +
  ylim(0, 100) + 
  coord_flip()
``` 

## Cut rates  

```{r plot cut percentage, echo=FALSE, fig.height=4, fig.width=14, message=F, warning=FALSE}
config$Reads_noPD <- config$Reads - config$PRIMER_DIMER
config$cut_percentage <- config$Cut * 100/config$Reads_noPD
config$cut_percentage[is.nan(config$cut_percentage)] <- 0  

ggplot(data = config, aes(x = as.factor(guideRNA), y = cut_percentage, order = guideRNA, fill = guideRNA)) +
  geom_boxplot() +
  xlab('guideRNA') + 
  ylab('Percentage of reads (not marked as PRIMER DIMERS) that have cut site')  +
  theme(axis.text = element_text(size=12),
        axis.title = element_text(size=14, face = 'bold'),
        legend.position = 'None') +
  ylim(0,100) + 
  coord_flip()
``` 

## Frameshift  
```{r plot frame shift percentage, echo=FALSE, fig.height=4, fig.width=14, message=F, warning=FALSE}
config$frameshift_percentage <- config$Frameshift * 100/config$Reads_noPD
config$frameshift_percentage[is.nan(config$frameshift_percentage)] <- 0  

ggplot(data = config, aes(x = as.factor(guideRNA), y = frameshift_percentage, order = guideRNA, fill = guideRNA)) +
  geom_boxplot() +
  xlab('guideRNA') + 
  ylab('Percentage of reads (not marked as PRIMER DIMERS) that have frameshift')  +
  theme(axis.text = element_text(size=12),
        axis.title = element_text(size=14,face = 'bold'),
        legend.position = 'None') +
  ylim(0, 100) + 
  coord_flip()
``` 

## Heterogeneity of reads  

```{r plot read heterogeneity, echo=FALSE, fig.height=4, fig.width=14, message=F, warning=FALSE}
alignments$ID_read_id <- paste0(alignments$seqnames, '_', alignments$read_id)
uniqueReadsByID <- alignments[!duplicated(alignments$ID_read_id), c('seqnames', 'read_id', 'count')]
uniqueReadsByID <- uniqueReadsByID[order(uniqueReadsByID$seqnames, uniqueReadsByID$count, decreasing = T),]

howManyTimes <- aggregate(read_id ~ seqnames, data = uniqueReadsByID, length)
howManyTimes <- howManyTimes[match(howManyTimes$seqnames, unique(uniqueReadsByID$seqnames)),] 

uniqueReadsByID$guideRNA <- rep(config$guideRNA[match(howManyTimes$seqnames, unique(config$ID))], 

                                times = howManyTimes$read_id)

uniqueReadsByID <- uniqueReadsByID[order(uniqueReadsByID$guideRNA, uniqueReadsByID$count, decreasing = T),]

cumsum_list <- tapply(uniqueReadsByID$count, as.vector(uniqueReadsByID$guideRNA), FUN = cumsum)
cumsum_list <- cumsum_list[match(names(cumsum_list), unique(uniqueReadsByID$guideRNA))]
uniqueReadsByID$cumsum <- do.call(c, cumsum_list)
seq2 <- Vectorize(seq.default, vectorize.args = c('from', 'to'))
howManyTimes <- table(as.vector(uniqueReadsByID$guideRNA))
howManyTimes <- howManyTimes[match(names(howManyTimes), unique(uniqueReadsByID$guideRNA))]
dimHMT <- if (is.null(dim(howManyTimes)[1])) { length(howManyTimes) } else { dim(howManyTimes)[1] }
uniqueReadsByID$read_number <- as.vector(unlist(seq2(from = rep(1, dimHMT), to = howManyTimes, by = 1)))

ids_with_reads <- tapply(uniqueReadsByID$cumsum, as.vector(uniqueReadsByID$guideRNA), FUN = max)
ids_with_reads <- ids_with_reads[match(names(ids_with_reads), unique(uniqueReadsByID$guideRNA))]

toDivide <- rep(ids_with_reads, times = howManyTimes)
uniqueReadsByID$read_share_percentage <- uniqueReadsByID$cumsum * 100 / toDivide
uniqueReadsByID$read_share_percentage_normal <- uniqueReadsByID$count * 100 / toDivide  

# divide into bins for colour
uniqueReadsByID$bins <- cut(uniqueReadsByID$read_share_percentage_normal,
                            c(0, 5, seq(10, 100, 10)))
# reduce number of reads in 0-5 group - faster plots without artifacts
uniqueReadsByID <- aggregate(read_share_percentage_normal ~ guideRNA + bins,
                             data = uniqueReadsByID, sum)
colorPalette <- colorRampPalette(c('#000000', '#F0E442'))(length(levels(uniqueReadsByID$bins)))
names(colorPalette) <- levels(uniqueReadsByID$bins)
ggplot(data = uniqueReadsByID, aes(x = as.factor(guideRNA), y = read_share_percentage_normal, fill = bins)) +
  geom_bar(position = 'stack', stat = 'identity') +
  theme(legend.position = 'top',
        axis.text = element_text(size=12),
        axis.title=element_text(size=14,face = 'bold'),
        legend.direction = 'horizontal', 
        legend.title = element_blank()) +
  ylab('Unique reads percentage of contribution') +
  xlab('guideRNA') +
  scale_fill_manual(values = colorPalette) +
  coord_flip()
``` 

***

# Alignments plots  

***

## AGGTGGTCAGGGAACTGG  

### Deletions  

```{r deletions AGGTGGTCAGGGAACTGG, echo=F, message=F, warning=F}
metaplot_deletions(alignments, config, 'guideRNA', c('AGGTGGTCAGGGAACTGG'))
```  

### Insertions  

```{r insertions AGGTGGTCAGGGAACTGG, echo=F, message=F, warning=F}
metaplot_insertions(alignments, config, 'guideRNA', c('AGGTGGTCAGGGAACTGG'))
```  

### Mismatches  

```{r mismatches AGGTGGTCAGGGAACTGG, echo=F, message=F, warning=F}
metaplot_mismatches(alignments, config, 'guideRNA', c('AGGTGGTCAGGGAACTGG'))
```  

## TGACCCTCTGCCAACACAAGGGG  

### Deletions  

```{r deletions TGACCCTCTGCCAACACAAGGGG, echo=F, message=F, warning=F}
metaplot_deletions(alignments, config, 'guideRNA', c('TGACCCTCTGCCAACACAAGGGG'))
```  

### Insertions  

```{r insertions TGACCCTCTGCCAACACAAGGGG, echo=F, message=F, warning=F}
metaplot_insertions(alignments, config, 'guideRNA', c('TGACCCTCTGCCAACACAAGGGG'))
```  

### Mismatches  

```{r mismatches TGACCCTCTGCCAACACAAGGGG, echo=F, message=F, warning=F}
metaplot_mismatches(alignments, config, 'guideRNA', c('TGACCCTCTGCCAACACAAGGGG'))
```  

## GTCCCTGCAACATTAAAGGCCGG  

### Deletions  

```{r deletions GTCCCTGCAACATTAAAGGCCGG, echo=F, message=F, warning=F}
metaplot_deletions(alignments, config, 'guideRNA', c('GTCCCTGCAACATTAAAGGCCGG'))
```  

### Insertions  

```{r insertions GTCCCTGCAACATTAAAGGCCGG, echo=F, message=F, warning=F}
metaplot_insertions(alignments, config, 'guideRNA', c('GTCCCTGCAACATTAAAGGCCGG'))
```  

### Mismatches  

```{r mismatches GTCCCTGCAACATTAAAGGCCGG, echo=F, message=F, warning=F}
metaplot_mismatches(alignments, config, 'guideRNA', c('GTCCCTGCAACATTAAAGGCCGG'))
```  

