---
title: 'Report breakdown by Amplicon'
author: amplican
date: '`r format(Sys.time(), "%d %B %Y")`'
output:
  html_document:
    toc: true
    theme: paper
    toc_float: true
    number_sections: true
---

```{r load data, message=F, warning=FALSE, include=FALSE}
results_folder = system.file('extdata', 'results', package = 'amplican')
library(amplican)
library(ggplot2)
alignments <- read.csv(file.path(results_folder, 'alignments_events.csv'), stringsAsFactors = FALSE)
config <- read.csv(file.path(results_folder, 'config_summary.csv'), stringsAsFactors = FALSE)
config$AmpliconUpper <- toupper(config$Amplicon)
uniqueAmlicons <- unique(config$AmpliconUpper)
labels <- sapply(uniqueAmlicons, function(x) {toString(config$ID[config$AmpliconUpper == x])})
```

***

# Description  

***

**Read distribution plot** - plot shows number of reads assigned during read grouping  
**PRIMER DIMER filter** - plot shows percentage of assigned reads that have been recognized as PRIMER DIMERS  
**Cutting rates** - plot gives overview of percentage of reads (not filtered as PRIMER DIMER) that have cut  
**Frameshift** - plot shows what percentage of reads that have frameshift  
**Read heterogeneity plot** - shows what is the share of each of the unique reads in total count of all reads. The more yellow each row, the less heterogeneity in the reads, more black means reads don't repeat often and are unique  
**Deletions plot** - shows summary of deletions detected after alignments with distinction
for forward (top plot) and reverse (bottom) reads, blue dotted lines represent primers as black
dotted line represents cut site box, for deletions overlapping with cut site box there is distinction
in color  
**Mismatches plot** - shows summary of mismatches detected after alignments split by forward
(top plot) and reverse (bottom) reads, mismatches are colored in the same manner as amplicon  
**Insertions plot** - shows summary of insertions detected after alignments split by forward
(top plot) and reverse (bottom) reads, insertion is shown as right-angled triangle where size of
the insertion corresponds to the width of the triangle, size and transparency of triangle reflect on
the frequency of the insertion  
**Cuts plot** - shows summary of cuts in the amplicon (insertions that are overlapping with 
specified box of uppercase letters in the amplicon, and are supported by both forward and reverse reads)

***

# Amplicon Summary  

***

## Amplicon groups

```{r amplicon groups, echo=FALSE, fig.height=4, fig.width=14, message=F, warning=FALSE}
library(knitr)
ampliconDF <- data.frame(group = 1:length(uniqueAmlicons), IDs = unname(labels))
kable(ampliconDF)
ampliconDF$Amplicon <- uniqueAmlicons #for sorting
config$group <- match(config$AmpliconUpper, ampliconDF$Amplicon)
```

## Read distribution  

```{r plot total reads, echo=FALSE, fig.height=4, fig.width=14, message=F, warning=FALSE}
ampliconTable <- aggregate(cbind(Reads, PRIMER_DIMER, Cut, Frameshift) ~ group, data = config, sum)

ggplot(data = ampliconTable, aes(x = as.factor(group), y = log10(Reads + 1)), order = group) +
  geom_bar(stat = 'identity') +
  ylab('Number of reads + 1, log10 scaled')  +
  xlab('Amplicon Group') +
  theme(legend.position = 'none',
        axis.text = element_text(size = 12),
        axis.title = element_text(size = 14, face = 'bold')) +
  coord_flip() +
  geom_text(aes(x = as.factor(group), y = log10(Reads + 1), label = Reads), hjust = -1)
```

## PRIMER DIMER filter  

```{r plot PRIMER DIMER percentage, echo=FALSE, fig.height=4, fig.width=14, message=F, warning=FALSE}

ampliconTable$PD_percentage <- ampliconTable$PRIMER_DIMER * 100/ampliconTable$Reads
ampliconTable$PD_percentage[is.nan(ampliconTable$PD_percentage)] <- 0 

ggplot(data = ampliconTable, aes(x = as.factor(group), y = PD_percentage, order = group)) +
  geom_bar(stat='identity') +
  ylab('Percentage of reads marked as PRIMER DIMERS')  +
  xlab('Amplicon Group') +
  theme(axis.text = element_text(size=12),
        axis.title = element_text(size=14, face = 'bold')) +
  ylim(0, 100) +
  coord_flip() +
  geom_text(aes(x = as.factor(group), y = PD_percentage, label = PRIMER_DIMER), hjust = -1)
```  

## Cut rates  

```{r plot cut rate, echo=FALSE, fig.height=4, fig.width=14, message=F, warning=FALSE}
ampliconTable$Reads_noPD <- ampliconTable$Reads - ampliconTable$PRIMER_DIMER
ampliconTable$cut_percentage <- ampliconTable$Cut * 100/ampliconTable$Reads_noPD
ampliconTable$cut_percentage[is.nan(ampliconTable$cut_percentage)] <- 0  

ggplot(data = ampliconTable, aes(x = as.factor(group), y = cut_percentage, order = group)) +
  geom_bar(stat = 'identity') +
  ylab('Percentage of reads (not marked as PRIMER DIMERS) that have cut site')  +
  xlab('Amplicon Group') +
  theme(axis.text = element_text(size=12),
        axis.title = element_text(size=14, face = 'bold')) +
  ylim(0,100) +
  coord_flip() +
  geom_text(aes(x = as.factor(group), y = cut_percentage, label = Cut), hjust = -1)
```  

## Frameshift  

```{r plot frame shift percentage, echo=FALSE, fig.height=4, fig.width=14, message=F, warning=FALSE}
ampliconTable$frameshift_percentage <- ampliconTable$Frameshift * 100/ampliconTable$Reads_noPD
ampliconTable$frameshift_percentage[is.nan(ampliconTable$frameshift_percentage)] <- 0 

ggplot(data = ampliconTable, aes(x = as.factor(group), y = frameshift_percentage, order = group)) +
  geom_bar(position = 'stack', stat = 'identity') +
  ylab('Percentage of reads (not marked as PRIMER DIMERS) that have frameshift')  +
  xlab('Amplicon Group') +
  theme(axis.text = element_text(size=12),
        axis.title = element_text(size=14,face = 'bold')) +
  ylim(0, 100) +
  coord_flip() +
  geom_text(aes(x = as.factor(group), y = frameshift_percentage, label = Frameshift), hjust = -1)
```  

## Heterogeneity of reads  

```{r plot read heterogeneity, echo=FALSE, fig.height=4, fig.width=14, message=F, warning=FALSE}
alignments$ID_read_id <- paste0(alignments$seqnames, '_', alignments$read_id)
uniqueReadsByID <- alignments[!duplicated(alignments$ID_read_id), c('seqnames', 'read_id', 'count')]
uniqueReadsByID <- uniqueReadsByID[order(uniqueReadsByID$seqnames, uniqueReadsByID$count, decreasing = T),]

howManyTimes <- aggregate(read_id ~ seqnames, data = uniqueReadsByID, length)
howManyTimes <- howManyTimes[match(howManyTimes$seqnames, unique(uniqueReadsByID$seqnames)),] 

uniqueReadsByID$group <- rep(config$group[match(howManyTimes$seqnames, unique(config$ID))], 
                             times = howManyTimes$read_id)

uniqueReadsByID <- uniqueReadsByID[order(uniqueReadsByID$group, uniqueReadsByID$count, decreasing = T),]

cumsum_list <- tapply(uniqueReadsByID$count, as.vector(uniqueReadsByID$group), FUN = cumsum)
cumsum_list <- cumsum_list[match(names(cumsum_list), unique(uniqueReadsByID$group))]
uniqueReadsByID$cumsum <- do.call(c, cumsum_list)

seq2 <- Vectorize(seq.default, vectorize.args = c('from', 'to'))
howManyTimes <- table(as.vector(uniqueReadsByID$group))
howManyTimes <- howManyTimes[match(names(howManyTimes), unique(uniqueReadsByID$group))]
dimHMT <- if (is.null(dim(howManyTimes)[1])) { length(howManyTimes) } else { dim(howManyTimes)[1] }
uniqueReadsByID$read_number <- as.vector(unlist(seq2(from = rep(1, dimHMT), to = howManyTimes, by = 1)))

ids_with_reads <- tapply(uniqueReadsByID$cumsum, as.vector(uniqueReadsByID$group), FUN = max)
ids_with_reads <- ids_with_reads[match(names(ids_with_reads), unique(uniqueReadsByID$group))]

toDivide <- rep(ids_with_reads, times = howManyTimes)
uniqueReadsByID$read_share_percentage <- uniqueReadsByID$cumsum * 100 / toDivide
uniqueReadsByID$read_share_percentage_normal <- uniqueReadsByID$count * 100 / toDivide  

# divide into bins for colour
uniqueReadsByID$bins <- cut(uniqueReadsByID$read_share_percentage_normal, 
                            c(0, 5, seq(10, 100, 10)))
# reduce number of reads in 0-5 group - faster plots without artifacts
uniqueReadsByID <- aggregate(read_share_percentage_normal ~ group + bins,
                             data = uniqueReadsByID, sum)
colorPalette <- colorRampPalette(c('#000000', '#F0E442'))(length(levels(uniqueReadsByID$bins)))
names(colorPalette) <- levels(uniqueReadsByID$bins)
ggplot(data = uniqueReadsByID, aes(x = as.factor(group), 
                                   y = read_share_percentage_normal, 
                                   fill = bins), 
       order = group) +
  geom_bar(position = 'stack', stat = 'identity') +
  theme(legend.position = 'top',
        axis.text = element_text(size=12),
        axis.title=element_text(size=14,face = 'bold'),
        legend.direction = 'horizontal', 
        legend.title = element_blank()) +
  ylab('Unique reads percentage of contribution') +
  xlab('Amplicon Group') +
  scale_fill_manual(values = colorPalette) +
  coord_flip()

# fix frequency to be relative to amplicon - not ID
alignments$group <- config$group[match(alignments$seqnames, config$ID)]
alignments$frequency <- alignments$count / ampliconTable$Reads[alignments$group]
``` 

# Alignments plots  

***

## Group 1  

### Deletions  

```{r deletions Group 1, echo=F, fig.height=14, fig.width=30, message=F, warning=F}
plot_deletions(alignments, config, c('ID_1','ID_3'), 5)
```  

### Insertions  

```{r insertions Group 1, echo=F, fig.height=14, fig.width=30, message=F, warning=F}
plot_insertions(alignments, config, c('ID_1','ID_3'), 5)
```  

### Mismatches  

```{r mismatches Group 1, echo=F, fig.height=14, fig.width=30, message=F, warning=F}
plot_mismatches(alignments, config, c('ID_1','ID_3'), 5)
```  

### Cuts  

```{r cuts Group 1, echo=F, fig.height=14, fig.width=30, message=F, warning=F}
plot_cuts(alignments, config, c('ID_1','ID_3'))
```  

## Group 2  

### Deletions  

```{r deletions Group 2, echo=F, fig.height=14, fig.width=30, message=F, warning=F}
plot_deletions(alignments, config, c('ID_2'), 5)
```  

### Insertions  

```{r insertions Group 2, echo=F, fig.height=14, fig.width=30, message=F, warning=F}
plot_insertions(alignments, config, c('ID_2'), 5)
```  

### Mismatches  

```{r mismatches Group 2, echo=F, fig.height=14, fig.width=30, message=F, warning=F}
plot_mismatches(alignments, config, c('ID_2'), 5)
```  

### Cuts  

```{r cuts Group 2, echo=F, fig.height=14, fig.width=30, message=F, warning=F}
plot_cuts(alignments, config, c('ID_2'))
```  

## Group 3  

### Deletions  

```{r deletions Group 3, echo=F, fig.height=14, fig.width=30, message=F, warning=F}
plot_deletions(alignments, config, c('ID_4'), 5)
```  

### Insertions  

```{r insertions Group 3, echo=F, fig.height=14, fig.width=30, message=F, warning=F}
plot_insertions(alignments, config, c('ID_4'), 5)
```  

### Mismatches  

```{r mismatches Group 3, echo=F, fig.height=14, fig.width=30, message=F, warning=F}
plot_mismatches(alignments, config, c('ID_4'), 5)
```  

### Cuts  

```{r cuts Group 3, echo=F, fig.height=14, fig.width=30, message=F, warning=F}
plot_cuts(alignments, config, c('ID_4'))
```  

