---
title: 'Report breakdown by ID'
author: amplican
date: '`r format(Sys.time(), "%d %B %Y")`'
output:
  html_document:
    toc: true
    theme: paper
    toc_float: true
    number_sections: true
---

```{r load data, message=F, warning=FALSE, include=FALSE}
results_folder = system.file('extdata', 'results', package = 'amplican')
library(amplican)
library(ggplot2)
alignments <- read.csv(file.path(results_folder, 'alignments_events.csv'), stringsAsFactors = FALSE)
config <- read.csv(file.path(results_folder, 'config_summary.csv'), stringsAsFactors = FALSE)
```

***

# Description  

***

**Read distribution plot** - plot shows number of reads assigned during read grouping  
**PRIMER DIMER filter** - plot shows percentage of assigned reads that have been recognized as PRIMER DIMERS  
**Cutting rates** - plot gives overview of percentage of reads (not filtered as PRIMER DIMER) that have cut  
**Frameshift** - plot shows what percentage of reads that have frameshift  
**Read heterogeneity plot** - shows what is the share of each of the unique reads in total count of all reads. The more blue each row, the less heterogeneity in the reads, more green means reads don't repeat often and are unique  
**Deletions plot** - shows summary of deletions detected after alignments with distinction
for forward (top plot) and reverse (bottom) reads, blue dotted lines represent primers as black
dotted line represents cut site box, for deletions overlapping with cut site box there is distinction
in color  
**Mismatches plot** - shows summary of mismatches detected after alignments split by forward
(top plot) and reverse (bottom) reads, mismatches are colored in the same manner as amplicon  
**Insertions plot** - shows summary of insertions detected after alignments split by forward
(top plot) and reverse (bottom) reads, insertion is shown as right-angled triangle where size of
the insertion corresponds to the width of the triangle, size and transparency of triangle reflect on
the frequency of the insertion

***

# ID Summary  

***

## Read distribution  

```{r plot total reads, echo=FALSE, fig.height=5, fig.width=14, message=F, warning=FALSE}
ggplot(data = config, aes(x = ID, y = log10(Reads + 1), order = ID)) +
  geom_bar(stat='identity') +
  ylab('Number of reads + 1, log10 scaled')  +
  theme(legend.position = 'none',
        axis.text = element_text(size = 12),
        axis.title = element_text(size = 14, face = 'bold')) +
  coord_flip() +
  geom_text(aes(x = ID, y = log10(Reads + 1), label = Reads), hjust = -1)
```

## PRIMER DIMER filter  

```{r plot PRIMER DIMER percentage, echo=FALSE, fig.height=5, fig.width=14, message=F, warning=FALSE}
config$PD_percentage <- config$PRIMER_DIMER * 100/config$Reads
config$PD_percentage[is.nan(config$PD_percentage)] <- 0  

ggplot(data = config, aes(x = ID, y = PD_percentage, order = ID)) +
  geom_bar(stat='identity') +
  ylab('Percentage of reads marked as PRIMER DIMERS')  +
  theme(legend.position = 'none',
        axis.text = element_text(size = 12),
        axis.title = element_text(size = 14, face = 'bold')) +
  coord_flip() +
  geom_text(aes(x = ID, y = PD_percentage, label = PRIMER_DIMER), hjust = -1)
```  

## Cut rates  

```{r plot cut percentage, echo=FALSE, fig.height=5, fig.width=14, message=F, warning=FALSE}
config$Reads_noPD <- config$Reads - config$PRIMER_DIMER
config$cut_percentage <- config$Cut * 100/config$Reads_noPD
config$cut_percentage[is.nan(config$cut_percentage)] <- 0  

ggplot(data = config, aes(x = ID, y = cut_percentage, order = ID)) +
  geom_bar(stat='identity') +
  ylab('Percentage of reads (not marked as PRIMER DIMERS) that have cut site')  +
  theme(legend.position = 'none',
        axis.text = element_text(size = 12),
        axis.title = element_text(size = 14, face = 'bold')) +
  coord_flip() +
  geom_text(aes(x = ID, y = cut_percentage, label = Cut), hjust = -1)
```  

## Frameshift  

```{r plot frame shift percentage, echo=FALSE, fig.height=5, fig.width=14, message=F, warning=FALSE}
config$frameshift_percentage <- config$Frameshift * 100/config$Reads_noPD
config$frameshift_percentage[is.nan(config$frameshift_percentage)] <- 0  

ggplot(data = config, aes(x = ID, y = frameshift_percentage, order = ID)) +
  geom_bar(stat='identity') +
  ylab('Percentage of reads (not marked as PRIMER DIMERS) that have frameshift')  +
  theme(legend.position = 'none',
        axis.text = element_text(size = 12),
        axis.title = element_text(size = 14, face = 'bold')) +
  coord_flip() +
  geom_text(aes(x = ID, y = frameshift_percentage, label = Frameshift), hjust = -1)
```  

## Heterogeneity of reads  

```{r plot read domination, echo=FALSE, fig.height=5, fig.width=14, message=F, warning=FALSE}
alignments$ID_read_id <- paste0(alignments$seqnames, '_', alignments$read_id)
uniqueReadsByID <- alignments[!duplicated(alignments$ID_read_id), c('seqnames', 'read_id', 'count')]
uniqueReadsByID <- uniqueReadsByID[order(uniqueReadsByID$seqnames, uniqueReadsByID$count, decreasing = T),]  

cumsum_list <- tapply(uniqueReadsByID$count, uniqueReadsByID$seqnames, FUN = cumsum)
cumsum_list <- cumsum_list[match(names(cumsum_list), unique(uniqueReadsByID$seqnames))]
uniqueReadsByID$cumsum <- do.call(c, cumsum_list)

howManyTimes <- aggregate(read_id ~ seqnames, data = uniqueReadsByID, length)

howManyTimes <- howManyTimes[match(howManyTimes$seqnames, unique(uniqueReadsByID$seqnames)),]  

seq2 <- Vectorize(seq.default, vectorize.args = c('from', 'to'))
uniqueReadsByID$read_number <- as.vector(unlist(seq2(from = rep(1, dim(howManyTimes)[1]), to = howManyTimes$read_id, by = 1)))  

ids_with_reads <- tapply(uniqueReadsByID$cumsum, uniqueReadsByID$seqnames, FUN = max)
ids_with_reads <- ids_with_reads[match(names(ids_with_reads), unique(uniqueReadsByID$seqnames))]
#config[match(howManyTimes$seqnames, config$ID),]
toDivide <- rep(ids_with_reads, times = howManyTimes$read_id)
uniqueReadsByID$read_share_percentage <- uniqueReadsByID$cumsum * 100 / toDivide
uniqueReadsByID$read_share_percentage_normal <- uniqueReadsByID$count * 100 / toDivide  

# divide into bins for colour
uniqueReadsByID$bins <- cut(uniqueReadsByID$read_share_percentage_normal,
                            c(0, 5, seq(10, 100, 10)))
# reduce number of reads in 0-5 group - faster plots without artifacts
uniqueReadsByID <- aggregate(read_share_percentage_normal ~ seqnames + bins,
                             data = uniqueReadsByID, sum)
colorPalette <- colorRampPalette(c('#009E73', '#0072B2'))(length(levels(uniqueReadsByID$bins)))
names(colorPalette) <- levels(uniqueReadsByID$bins)
ggplot(data = uniqueReadsByID, aes(x = seqnames, y = read_share_percentage_normal, fill = bins)) +
  geom_bar(position='stack', stat='identity') +
  theme(axis.text = element_text(size = 12),
        axis.title = element_text(size = 14, face = 'bold'),
        legend.position = 'top',
        legend.direction = 'horizontal',
        legend.title = element_blank()) +
  ylab('Unique reads percentage of shares') +
  xlab('ID') +
  scale_fill_manual(values = colorPalette) +
  coord_flip()
```  

***

# Alignments plots  

***

## ID_1  

### Deletions  

```{r deletions ID_1, echo=F, fig.height=14, fig.width=30, message=F, warning=F}
amplican_plot_deletions(alignments, config, c('ID_1'), 5)
```  

### Insertions  

```{r insertions ID_1, echo=F, fig.height=14, fig.width=30, message=F, warning=F}
amplican_plot_insertions(alignments, config, c('ID_1'), 5)
```  

### Mismatches  

```{r mismatches ID_1, echo=F, fig.height=14, fig.width=30, message=F, warning=F}
amplican_plot_mismatches(alignments, config, c('ID_1'), 5)
```  

## ID_2  

### Deletions  

```{r deletions ID_2, echo=F, fig.height=14, fig.width=30, message=F, warning=F}
amplican_plot_deletions(alignments, config, c('ID_2'), 5)
```  

### Insertions  

```{r insertions ID_2, echo=F, fig.height=14, fig.width=30, message=F, warning=F}
amplican_plot_insertions(alignments, config, c('ID_2'), 5)
```  

### Mismatches  

```{r mismatches ID_2, echo=F, fig.height=14, fig.width=30, message=F, warning=F}
amplican_plot_mismatches(alignments, config, c('ID_2'), 5)
```  

## ID_3  

### Deletions  

```{r deletions ID_3, echo=F, fig.height=14, fig.width=30, message=F, warning=F}
amplican_plot_deletions(alignments, config, c('ID_3'), 5)
```  

### Insertions  

```{r insertions ID_3, echo=F, fig.height=14, fig.width=30, message=F, warning=F}
amplican_plot_insertions(alignments, config, c('ID_3'), 5)
```  

### Mismatches  

```{r mismatches ID_3, echo=F, fig.height=14, fig.width=30, message=F, warning=F}
amplican_plot_mismatches(alignments, config, c('ID_3'), 5)
```  

## ID_4  

### Deletions  

```{r deletions ID_4, echo=F, fig.height=14, fig.width=30, message=F, warning=F}
amplican_plot_deletions(alignments, config, c('ID_4'), 5)
```  

### Insertions  

```{r insertions ID_4, echo=F, fig.height=14, fig.width=30, message=F, warning=F}
amplican_plot_insertions(alignments, config, c('ID_4'), 5)
```  

### Mismatches  

```{r mismatches ID_4, echo=F, fig.height=14, fig.width=30, message=F, warning=F}
amplican_plot_mismatches(alignments, config, c('ID_4'), 5)
```  

